add support for `ord` = 2, -2, and 'nuc' in `cupy.linalg.norm`

rushabh-v opened the pull request:
closes #3053

leofang commented:
Hi @rushabh-v, thank you for providing better support! I haven't read your implementation carefully, but I think you need to add a few tests covering the newly supported `ord` to here:
https://github.com/cupy/cupy/blob/9cfbed5d2f50191ec8f990dee5c1ae5ebae7167b/tests/cupy_tests/linalg_tests/test_norms.py#L49
I'll be more than happy to assist you if you need any feedback.

leofang commented on cupy/linalg/norms.py:
```diff
@@ -6,6 +6,30 @@
+        The return values are either the minimum or maximum or sum of the
+        singular values of the matrices, depending on whether `op`
+        is `cupy.amin` or `cupy.amax` or `cupy.sum`.
+    """
```
I'd suggest removing the entire docstring for a few reasons:
1. This is only a private function and not exposed to end users
2. This helper function is simple enough (only two lines)
3. The docstring format does not match the CuPy (or Google) style

For the docstring style, you can take a look at `norm()`'s docstring right below for how to properly format it.

rushabh-v replied:
I have removed that docstring.
And I will take care of that docstring convention next time onwards : )

rushabh-v commented:
Hey @leofang . Thanks for your response.

I actually tried adding the tests. And if I understood correctly, the tests for `ord=2` and `ord=-2` are already there but the errors were ignored. see, https://github.com/cupy/cupy/blob/9cfbed5d2f50191ec8f990dee5c1ae5ebae7167b/tests/cupy_tests/linalg_tests/test_norms.py#L28

Correct me if I am wrong at this. Because I am still trying to understand the testing module of CuPy properly as I found a bit different and advanced than other libraries have contributed to.

asi1024 commented:
@rushabh-v Thanks for your PR! Could you add tests in `tests/cupy_tests/linalg_tests/test_norms.py`?

leofang commented:
> the tests for ord=2 and ord=-2 are already there but the errors were ignored

No, the error is not ignored. I think the purpose of `testing.NumpyError` is to ignore the division error only for backward compatibility with older NumPy. (I think you can try removing that line actually.) 

You need to ensure the case reported in #3053 is covered, which means you need to test against 2D matrices, which means you need to add `ord=2,-2` to this line:
https://github.com/cupy/cupy/blob/9cfbed5d2f50191ec8f990dee5c1ae5ebae7167b/tests/cupy_tests/linalg_tests/test_norms.py#L33
I have not tested extensively what's covered here and what's not (compared to NumPy), so for the moment I don't have further comment.

rushabh-v commented:
Oh yes, I mistakenly didn't focus on the shape. I have added the tests now.

leofang commented:
Have you tried running the tests locally? You can do 
```python
python -m pytest -v tests/cupy_tests/linalg_tests/test_norms.py
```
(or replace `pytest` by `unittest`).

rushabh-v commented:
> Have you tried running the tests locally? 

Yes, I have tested it locally. It works fine.

toslunar commented on cupy/linalg/norms.py:
```diff
@@ -97,7 +103,11 @@ def norm(x, ord=None, axis=None, keepdims=False):
+        if ord == 2:
+            ret = _multi_svd_norm(x, row_axis, col_axis, cupy.amax)
+        elif ord == -2:
+            ret = _multi_svd_norm(x, row_axis, col_axis, cupy.amin)
```
SVD returns the singular values in descending order.  `functools.partial(cupy.take, indices=0)` or `functools.partial(cupy.take, indices=-1)` could be used, while SVD is much heavier than max/min.

rushabh-v replied:
I have added that.

leofang commented:
> Yes, I have tested it locally. It works fine.

Hmm. I applied your patch but It failed on my machine: 
```bash
E                   AssertionError: Only numpy raises error
E                   
E                     File "/home/leofang/test/cupy2/cupy/testing/helper.py", line 22, in _call_func
E                       result = impl(self, *args, **kw)
E                     File "/home/leofang/test/cupy2/tests/cupy_tests/linalg_tests/test_norms.py", line 58, in test_norm
E                       return xp.linalg.norm(a, self.ord, self.axis, self.keepdims)
E                     File "<__array_function__ internals>", line 6, in norm
E                     File "/home/leofang/conda_envs/from_CF/lib/python3.7/site-packages/numpy/linalg/linalg.py", line 2533, in norm
E                       ret = _multi_svd_norm(x, row_axis, col_axis, amin)
E                     File "/home/leofang/conda_envs/from_CF/lib/python3.7/site-packages/numpy/linalg/linalg.py", line 2317, in _multi_svd_norm
E                       result = op(svd(y, compute_uv=0), axis=-1)
E                     File "<__array_function__ internals>", line 6, in svd
E                     File "/home/leofang/conda_envs/from_CF/lib/python3.7/site-packages/numpy/linalg/linalg.py", line 1618, in svd
E                       t, result_t = _commonType(a)
E                     File "/home/leofang/conda_envs/from_CF/lib/python3.7/site-packages/numpy/linalg/linalg.py", line 156, in _commonType
E                       (a.dtype.name,))
E                   array type float16 is unsupported in linalg
```
Which NumPy version are you using? I'm on 1.17.2.

rushabh-v commented:
I actually don't have Nvidia GPU. So I don't have cupy installed.

I should not have said "Locally" there. What I did was I edited the file `norm.py` in google colab and checked whether both, numpy and cupy generate the same output or not.

I am wondering if there is a better way of doing this.

rushabh-v commented:
I think it's a bug there in NumPy. See https://github.com/numpy/numpy/issues/11690 
note: I have removed `with testing.NumpyError(divide='ignore'):`

rushabh-v commented:
@leofang I am not sure if it was because I removed the `with testing.NumpyError(divide='ignore'):`. If so, I have added that line back in the last commit. Can you test it back, please? 
I am sorry for the inconvenience and my mistake in the testing.

leofang commented:
Hi @rushabh-v GJ for finding the related issue! I even left a comment in the PR that would fix it a while ago 😅 

So, it seems all we need to do is just not to test against `float16` (because NumPy's SVD does not yet support it, as pointed out in the issue you linked). You can do this simply by changing the decorator `@testing.for_all_dtypes()` to `@testing.for_all_dtypes(no_float16=True)` for the test method.

I have not used colab for a long time, so I have no recommendation for the best practice using it as a develop env. Please do let me know if you figure it out!

You can still remove `@testing.NumpyError`. It doesn't matter.

rushabh-v commented:
I have added a commit. Can you check it, please? @leofang

rushabh-v commented:
> Please do let me know if you figure it out!

@leofang  I figured out a way to use colab as a development env. It's simply the way we use our local machine as a development env.
cloning the patch there, building that from source and then testing it.

Here is the script I used for my patch,
```
import os
!git clone https://github.com/rushabh-v/cupy
!cd cupy
os.chdir("cupy/")
!pip install -e . --no-cache-dir -vvvv
!git checkout ord-norm
!python3 -m pytest -v tests/cupy_tests/linalg_tests/test_norms.py
```

Found 100% of the tests passing. Tell me if it requires any changes now.

leofang commented:
Thanks for sharing, @rushabh-v. LGTM.

rushabh-v commented on cupy/linalg/norms.py:
```diff
@@ -5,6 +5,14 @@
 from cupy.linalg import decomposition
 from cupy.linalg import util
 
+from functools import partial
```
Okay!
Then in what another way can I do this? can you please tell me?

asi1024 replied:
Please use `functools.partial`.

rushabh-v replied:
Oh, okay. My bad!
I thought, I can't import functools also. Will be making that change soon.

rushabh-v replied:
Done

asi1024 commented:
Jenkins, test this please.

asi1024 commented:
LGTM. Thanks for your PR!

leofang commented:
Thanks again, @rushabh-v!

rushabh-v commented:
You're welcome : )
I am actually planning to apply to CuPy for GSoC this year. And I am interested to work on "CuPy coverage of NumPy functions" or "Support for New Random Generator interface".

So you guys can tell me if there is any potential contribution I can make. I will be absolutely available.
